File: scripts/check-architecture.mjs
import fs from 'node:fs';
import path from 'node:path';
import { spawnSync } from 'node:child_process';
import { fileURLToPath } from 'node:url';
import ts from 'typescript';

export function sourceFiles(directory) {
  return fs
    .readdirSync(directory, { withFileTypes: true })
    .flatMap((entry) =>
      entry.isDirectory()
        ? sourceFiles(`${directory}/${entry.name}`)
        : /\.(tsx?|css)$/.test(entry.name)
          ? [`${directory}/${entry.name}`]
          : [],
    );
}
export function checkFrontend(files) {
  const errors = [],
    graph = new Map();
  const feature = (file) => file.match(/^src\/features\/([^/]+)/)?.[1];
  function resolve(file, spec) {
    const base = path.posix.normalize(
      path.posix.join(path.posix.dirname(file), spec.split('?')[0]),
    );
    return [base, base + '.ts', base + '.tsx', base + '/index.ts', base + '/index.tsx'].find((p) =>
      files.has(p),
    );
  }
  for (const [file, source] of files) {
    const imports = file.endsWith('.css')
      ? [...source.matchAll(/@import\s+(?:url\()?['"]([^'"]+)['"]/g)]
          .map((match) => `import ${JSON.stringify(match[1])};`)
          .join('\n')
      : source;
    const ast = ts.createSourceFile(file, imports, ts.ScriptTarget.Latest, true),
      dependencies = [];
    const visit = (node) => {
      let spec,
        typeOnly = false;
      if (
        (ts.isImportDeclaration(node) || ts.isExportDeclaration(node)) &&
        node.moduleSpecifier &&
        ts.isStringLiteral(node.moduleSpecifier)
      ) {
        spec = node.moduleSpecifier.text;
        typeOnly = !!node.isTypeOnly || !!node.importClause?.isTypeOnly;
        if (
          ts.isImportDeclaration(node) &&
          node.importClause?.namedBindings &&
          ts.isNamedImports(node.importClause.namedBindings) &&
          !node.importClause.name
        )
          typeOnly ||= node.importClause.namedBindings.elements.every((n) => n.isTypeOnly);
      }
      if (
        ts.isCallExpression(node) &&
        node.expression.kind === ts.SyntaxKind.ImportKeyword &&
        ts.isStringLiteral(node.arguments[0])
      )
        spec = node.arguments[0].text;
      if (spec) {
        if (
          /\/lessons\/[^/]+\/(trace|presets)\.ts$/.test(file) &&
          /^(react|react-dom|@tanstack\/react-query)(\/|$)/.test(spec) &&
          !typeOnly
        )
          errors.push(`${file}: pure lesson code imports ${spec}`);
        if (spec.startsWith('.')) {
          const target = resolve(file, spec);
          if (target) {
            if (file.startsWith('src/shared/') && !target.startsWith('src/shared/'))
              errors.push(`${file}: shared code cannot depend on ${target}`);
            if (
              feature(file) &&
              feature(target) &&
              feature(file) !== feature(target) &&
              !target.endsWith('/index.ts') &&
              !target.endsWith('/index.tsx')
            )
              errors.push(`${file}: cross-feature import must use the public entry: ${target}`);
            if (file.startsWith('src/features/') && target.startsWith('src/app/'))
              errors.push(`${file}: features cannot depend on the application shell`);
            if (
              file.includes('/visualizers/core/') &&
              /\/visualizers\/(lessons|components|registry)/.test(target)
            )
              errors.push(`${file}: visualizer core depends on a lesson or renderer`);
            if (
              /\/lessons\/[^/]+\/trace\.ts$/.test(file) &&
              (/\.(tsx)$/.test(target) ||
                target.includes('/shared/api/') ||
                target.endsWith('/adapter.ts'))
            )
              errors.push(
                `${file}: trace generation must remain independent of presentation and API access`,
              );
            if (!typeOnly) dependencies.push(target);
          } else if (!spec.includes('?raw') && !/\.(py|png|svg)$/.test(spec))
            errors.push(`${file}: unresolved internal module ${spec}`);
        }
      }
      ts.forEachChild(node, visit);
    };
    visit(ast);
    graph.set(file, dependencies);
  }
  const visited = new Set(),
    active = [];
  function visit(file) {
    if (active.includes(file)) {
      errors.push(
        'Dependency cycle: ' + [...active.slice(active.indexOf(file)), file].join(' -> '),
      );
      return;
    }
    if (visited.has(file)) return;
    active.push(file);
    for (const next of graph.get(file) ?? []) visit(next);
    active.pop();
    visited.add(file);
  }
  for (const file of graph.keys()) visit(file);
  return errors;
}
if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  const errors = checkFrontend(
    new Map(sourceFiles('src').map((file) => [file, fs.readFileSync(file, 'utf8')])),
  );
  if (errors.length) {
    console.error(errors.join('\n'));
    process.exitCode = 1;
  } else
    console.log('Frontend architecture: boundaries, module resolution, and runtime cycles passed.');
  const python =
    process.env.PYTHON ??
    ['.venv/Scripts/python.exe', '.venv/bin/python'].find((p) => fs.existsSync(p)) ??
    (process.platform === 'win32' ? 'python' : 'python3');
  const result = spawnSync(python, ['scripts/check-architecture.py'], { stdio: 'inherit' });
  if (result.status !== 0) process.exitCode = 1;
}

File: scripts/check-architecture.py
"""Check backend layer direction and dependency cycles without third-party packages."""

import ast
from pathlib import Path


def check_backend(root: Path) -> list[str]:
    modules = {}
    for file in root.rglob("*.py"):
        relative = file.relative_to(root).with_suffix("")
        parts = list(relative.parts)
        if parts[-1] == "__init__":
            parts.pop()
        modules["algo_atlas" + ("." + ".".join(parts) if parts else "")] = file
    errors = []
    graph = {}
    for name, file in modules.items():
        package = name if file.name == "__init__.py" else name.rsplit(".", 1)[0]
        dependencies = []
        for node in ast.walk(ast.parse(file.read_text(encoding="utf-8"))):
            imports = []
            if isinstance(node, ast.Import):
                imports = [entry.name for entry in node.names]
            elif isinstance(node, ast.ImportFrom):
                base = node.module or ""
                if node.level:
                    prefix = package.split(".")[: len(package.split(".")) - node.level + 1]
                    base = ".".join(prefix + ([base] if base else []))
                imports = [
                    base + "." + entry.name if base + "." + entry.name in modules else base
                    for entry in node.names
                ]
            for target in imports:
                layer = name.split(".")[1] if "." in name else ""
                target_layer = target.split(".")[1] if target.startswith("algo_atlas.") else target
                forbidden = {
                    "api": {"integrations", "persistence"},
                    "services": {"api", "main", "fastapi"},
                    "persistence": {"api", "services", "integrations", "main", "bootstrap"},
                    "integrations": {"api", "services", "main", "bootstrap"},
                }
                if target_layer in forbidden.get(layer, set()):
                    errors.append(f"{name}: {layer} must not import {target}")
                if target in modules:
                    dependencies.append(target)
        graph[name] = dependencies
    visited, active = set(), []

    def visit(name):
        if name in active:
            errors.append("Dependency cycle: " + " -> ".join(active[active.index(name) :] + [name]))
            return
        if name in visited:
            return
        active.append(name)
        for dependency in graph.get(name, []):
            visit(dependency)
        active.pop()
        visited.add(name)

    for name in graph:
        visit(name)
    return errors


if __name__ == "__main__":
    issues = check_backend(Path("backend/algo_atlas"))
    if issues:
        raise SystemExit("\n".join(issues))
    print("Backend architecture: layer boundaries and dependency cycles passed.")
